File: src/2022/Day4/solution.ts
import fs from "fs/promises";
import { CheckSections } from "./types";

async function main(file: string) {
  const input: string = await fs.readFile(file, { encoding: "utf-8" });
  const pairs: number[][][] = input
    .split("\r\n")
    .map((pair) => pair.split(",").map((elf) => elf.split("-").map(Number)));

  return {
    PartA: getNumOfOverlaps(pairs, isFullyContained),
    PartB: getNumOfOverlaps(pairs, isOverlaping),
  };
}

function isFullyContained(sectionA: Array<number>, sectionB: Array<number>): boolean {
  return (
    (sectionA[0] >= sectionB[0] && sectionA[1] <= sectionB[1]) ||
    (sectionA[0] <= sectionB[0] && sectionA[1] >= sectionB[1])
  );
}

function isOverlaping(sectionA: Array<number>, sectionB: Array<number>): boolean {
  return sectionA[1] >= sectionB[0] && sectionB[1] >= sectionA[0];
}

function getNumOfOverlaps(sections: number[][][], checkFunc: CheckSections) {
  return sections.reduce((total, pair) => {
    const [sectionA, sectionB] = pair;
    const isOverlaping = checkFunc(sectionA, sectionB);
    return total + Number(isOverlaping);
  }, 0);
}

main("input.txt").then((res) => console.log(res));

File: src/2022/Day4/solution.py
with open("input.txt") as f:
    data = f.read()
data = data.replace("\r\n", "\n") # Just in case I run this in Windows
data = data.split("\n")

pairs = [[list(map(int, range_pair.split("-"))) for range_pair in line.split(",")] for line in data]

def is_fully_contained(section_a: list, section_b: list) -> bool:
    return (section_a[0] >= section_b[0] and section_a[1] <= section_b[1]) or (section_a[0] <= section_b[0] and section_a[1] >= section_b[1])

def is_overlaping(section_a: list, section_b: list) -> bool:
    return section_a[1] >= section_b[0] and section_b[1] >= section_a[0]

fully_contained_count = sum(is_fully_contained(pair[0], pair[1]) for pair in pairs)
overlaping_count = sum(is_overlaping(pair[0], pair[1]) for pair in pairs)

print("Part A: ", fully_contained_count)
print("Part B: ", overlaping_count)
